cli: clean up the code around the handling of default parameters

knz opened the pull request.

knz committed:
cli: move the definition of pflag Value types into a separate file

This highlights that they are meant to be reusable across commands,
and provides a single point for folk to go to if they are looking for
a pflag.Value when implementing a new feature.

Release note: None

knz committed:
cli: topologically sort the structs in context.go

Release note: none

knz committed:
cli: move maxResults to debugCtx

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20987)
<!-- Reviewable:end -->

knz committed:
cli: centralize the definition of configuration defaults

Prior to this patch, the default values for configuration parameters
were duplicated in `flags.go` and `TestingRest()` in
`testutils.go`. The latter was necessary because some tests to not go
through command-line parsing.

This patch captures the default initialization logic into a single
location (`initCLIDefaults()`), then:

- calls that from `TestingRest()`,
- calls that from the `init` function in `flags.go`,
- uses the (now pre-computed) defaults as default parameter for the
  `pflag` variable creation calls in `flags.go`.

Release note: None

knz commented:
We might want to cherry pick some (if not all) of this in 1.1. The faulty re-initialization of defaults in between tests has bitten us in the past.

knz committed:
cli: move cmdTimeout to cliCtx

Release note: None

knz committed:
cli: move tempDir and externalIODir to startCtx

Release note: None

knz committed:
cli: move the SQL connection parameters to cliCtx

Release note: None

knz committed:
cli: properly (re-)initialize some server defaults

Release note: None

knz committed:
cli: properly re-use some pre-computed default values

Release note: None

knz committed:
cli: avoid leaking a global var in flags.go out of flags.go

Release note: None

knz committed:
cli: centralize the default initialization of special vars in flags.go

Release note: None

knz commented:
@tschottdorf @mberhault perhaps one of you could have a look since alex is out?

mberhault commented:
I'll take a look. At first glance though I would say this doesn't go into 1.1 unless this fixes bugs in non-testing code.

mberhault commented on pkg/cli/flags.go:
```diff
@@ -20,36 +20,45 @@ import (
-
-var sqlConnURL, sqlConnUser, sqlConnDBName string
-var serverConnHost, serverConnPort, serverAdvertiseHost, serverAdvertisePort string
+// special global variables used by flag variable definitions belows.
```
s/belows/below/

mberhault commented on pkg/cli/flags_util.go:
```diff
@@ -0,0 +1,452 @@
+	"golang.org/x/net/context"
+)
+
+// This file contains definitions for data types suitable for use by
```
I don't think this whole section is necessary. Not only is `flag.Value` a subset of the `pflag.Value` interface, types that don't implement `pflag.Value` will spit out a clear error when using them on `VarFlag`.

mberhault reviewed (approved):
LGTM. Thanks for the refactor. Only small nits (that I don't feel strongly about).

No cherry-pick into 1.1 though. If we have specific flag issues in 1.1 we can fix them individually, but definitely not the whole refactor.

knz commented:
Thanks!

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/cli/flags.go, line 33 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/20987#-L0tg79KDifpislJ53at-r11-33:-L1NdJPyWCKF6Ix0v-CP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d101af6d3c1f0f1afe7584fb1629839cd23bad7/pkg/cli/flags.go#L33)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

s/belows/below/
</blockquote></details>

Done.

---

*[pkg/cli/flags.go, line 50 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/20987#-L0tg79KDifpislJ53at-r11-50:-L1NdJAUbTErpwraRwf-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d101af6d3c1f0f1afe7584fb1629839cd23bad7/pkg/cli/flags.go#L50)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

s/initialize/initializes/
</blockquote></details>

Done.

---

*[pkg/cli/flags_util.go, line 36 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/20987#-L0tg79JKBpTzqe0SqS_-r11-36:-L1NdLQjMdufxqZaDBj0:bhgwo7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d101af6d3c1f0f1afe7584fb1629839cd23bad7/pkg/cli/flags_util.go#L36)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I don't think this whole section is necessary. Not only is `flag.Value` a subset of the `pflag.Value` interface, types that don't implement `pflag.Value` will spit out a clear error when using them on `VarFlag`.
</blockquote></details>

Agreed. Removed. (The code was there in different places previously. This simplifies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20987#-:-L1Nexo4WNo1iNKpsoxV:b-ucfknh)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR :)
